### PR TITLE
refactor(tests): Rewrite achievement tests

### PR DESCRIPTION
**Pull request by Androbin**
### Problem
Reduce the amount of code that needs to be maintained for testing.

### Solution
Factor out common assertion patterns

### Areas of Impact
This affects most `.js` files directly in the `test/` directory and doesn't change any behaviour.

### CODE REVIEW

**Comment by Androbin** [comment]
@LordSputnik I am doing my best to modularize this as best as I can.

File: test/common.js [hunk 1]
```diff
@@ -45,3 +59,39 @@ export function expectAchievementIdsNested(promise, name,
...
+	return () => rewireAchievement({
+		getTypeCreation:
+			testData.typeCreationHelper(
+				`${name}_revision`, threshold
+			)
+	});
+}
+
+export function testAchievement(rewiring, generator, expectations) {
+	return () => {
```
Comment (LordSputnik) [inline test/common.js]: This probably shouldn't return a function. I'd expect it to just be:

    testAchievement(...) {
        rewiring();
        const promise = generator();
        expectations(promise);
    }

File: test/common.js [hunk 2]
```diff
@@ -45,3 +59,39 @@ export function expectAchievementIdsNested(promise, name,
...
+		)
+		.then((edit) => edit[func][name]);
+}
+
+export function rewireAchievement(rewiring) {
+	return () => Achievement.__set__(rewiring);
+}
+
+export function rewireTypeCreation(name, threshold) {
+	return () => rewireAchievement({
```
Comment (LordSputnik) [inline test/common.js]: This shouldn't return a function.

File: test/common.js [hunk 3]
```diff
@@ -45,3 +59,39 @@ export function expectAchievementIdsNested(promise, name,
...
+	return () => new Editor({name: testData.editorAttribs.name})
+		.fetch()
+		.then((editor) =>
+			Achievement.processEdit(orm, editor.id)
+		)
+		.then((edit) => edit[func][name]);
+}
+
+export function rewireAchievement(rewiring) {
+	return () => Achievement.__set__(rewiring);
```
Comment (LordSputnik) [inline test/common.js]: This shouldn't return a function. Also, better to not use globals, so I'd say pass in the `Achievement` object and then change it within the function (and go back to calling `const Achievement = rewire(...)` in each Achievement module.

There may be unintended side effects to having a shared Achievement variable (not being cleaned up properly after each test, for example).

File: test/common.js [hunk 4]
```diff
@@ -45,3 +59,39 @@ export function expectAchievementIdsNested(promise, name,
 			.property(`${name}.titleId`, titleId)
 	]);
 }
+
+export function generateProcessEdit(func, name, rev) {
+	return () => testData.createEditor()
```
Comment (LordSputnik) [inline test/common.js]: Shouldn't return a function.

File: test/common.js [hunk 5]
```diff
@@ -45,3 +59,39 @@ export function expectAchievementIdsNested(promise, name,
...
+		.then((editor) => Achievement.processEdit(orm, editor.id))
+		.then((edit) => edit[func][`${name} ${rev}`]);
+}
+
+export function generateProcessEditNamed(func, name) {
+	return () => new Editor({name: testData.editorAttribs.name})
```
Comment (LordSputnik) [inline test/common.js]: Shouldn't return a function.

**Comment by Androbin** [comment]
Since all those functions in `common.js` are to be passed as a function, I thought it would be less boilerplate after all.

File: test/common.js (hunk 3 above)
Comment (Androbin) [reply test/common.js]: What about changing this to use the `__with__` method?

File: test/common.js (hunk 1 above)
Comment (Androbin) [reply test/common.js]: > So, we'd like you to write a function that takes those as parameters, and either
> a) returns a function that'll test a particular achievement
> b) test the achievement itself

Comment (LordSputnik) [reply test/common.js]: OK, so if you want to take that approach, you'll need to do something like:

    const test1 = testAchievement(a, b, c); // at this point test1 is a function taking no parameters generated by testAchievement
    it('...', test1);

Comment (Androbin) [reply test/common.js]: Shall I disassemble the parameters in the same way?

**Comment by Androbin** [comment]
@LordSputnik Travis gives error messages I don't understand.

**Comment by Androbin** [comment]
@LordSputnik Can you tell what this error message means?
`Unhandled rejection error: duplicate key value violates unique constraint "editor_type_pkey"`

**Comment by Androbin** [comment]
Some quick research suggests that the database got "out of sync".

**Comment by Androbin** [comment]
@LordSputnik I have problems dealing with this CI error:
`AssertionError: expected false to have property 'editorId'`
I suspect that some value isn't passed properly all the way.

**Comment by Androbin** [comment]
@LordSputnik Failing `master` is blocking the CI tests on this PR.
PS. Yeah, I'm sorry, it fails to run locally.

**Comment by Androbin** [comment]
I am experimenting a bit with passing deps to common instead of importing from there.

**Comment by LordSputnik** [comment]
Hi @Androbin,

I think it would probably be quicker to start from scratch and make this change in small steps. Transform the code in one way per commit, check the tests work and then make the next step. If the tests don't run successfully locally with the current master, let me know and we can try to fix it.

**Comment by Androbin** [comment]
@LordSputnik Okay, I will split into more granular commits.

**Comment by Androbin** [comment]
I have parked the original patch set on [branch refactor](https://github.com/Androbin/bookbrainz-site/tree/refactor).

**Comment by Androbin** [comment]
Status Quo:
- `common.js` looks nearly like before.
- tests are split into parts again

Yet To Do:
- de-duplicate file-wise
- refactor more customizations

**Comment by Androbin** [comment]
Also, I'm thinking about passing `Achievement` and/or `orm` to `testAchievement`, not earlier.

**Comment by Androbin** [comment]
@LordSputnik Turns out that tests fail if the rewiring is reverted after each test.

**Comment by Androbin** [comment]
@LordSputnik Any objections?

**Pull request merged**